pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -461,11 +468,57 @@ func (c *kvEventToRowConsumer) encodeAndEmit(
 		return err
 	}
 	if log.V(3) {
-		log.Infof(ctx, `r %s: %s -> %s`, updatedRow.TableName, keyCopy, valueCopy)
+		log.Infof(ctx, `r %s(%+v): %s -> %s`, updatedRow.TableName, keyCopy, headers, valueCopy)
 	}
 	return nil
 }
 
+var jsonHeaderWrongTypeLogLim = log.Every(1 * time.Minute)
+
+func (c *kvEventToRowConsumer) makeRowHeaders(
+	ctx context.Context, updatedRow cdcevent.Row,
+) (map[string][]byte, error) {
+	if c.encodingOpts.HeadersJSONColName == "" {
+		return nil, nil
+	}
+
+	headersJSON, err := readOneJSONValue(updatedRow, c.encodingOpts.HeadersJSONColName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read headers column %s: %s", c.encodingOpts.HeadersJSONColName, updatedRow.DebugString())
+	}
+	// Allow & ignore both kinds of null.
+	if headersJSON == nil || headersJSON.Type() == json.NullJSONType {
+		return nil, nil
+	}
+
+	headers := make(map[string][]byte, headersJSON.Len())
+	objIt, err := headersJSON.ObjectIter()
+	if err != nil {
+		return nil, err
+	}
+	if objIt == nil {
+		if jsonHeaderWrongTypeLogLim.ShouldLog() || log.V(2) {
+			log.Warningf(ctx, "headers column %s must be a JSON object, got %s - %s", c.encodingOpts.HeadersJSONColName, headersJSON.Type(), headersJSON)
+		}
+		return nil, nil
+	}
+	// Format the object's kv pairs. The intended use case is to have string
+	// values, but permit other types, formatted as JSON. I'd rather have
+	// confusing headers than the feed choke on specific values.
+	for objIt.Next() {
+		var val string
+		switch objIt.Value().Type() {
+		case json.StringJSONType:
+			str, _ := objIt.Value().AsText()
+			val = *str
+		default:
+			val = objIt.Value().String()
+		}
+		headers[objIt.Key()] = []byte(val)
```

Comment:
Just confirming that because we're parsing the headers from json, we're not at risk of overriding an existing header with a different value, right? I.e., no duplicate keys.

Reply:
good question. i don't think duplicates can appear in jsonb types but i can add a test for it.

```
 select '{"a": 1, "a": 2}'::jsonb;
   jsonb
------------
  {"a": 2}
(1 row)
```